### ISSUE TITLE
Pls. reduce FeeHistory query range from 20.000 to 1024 for long term queries

### ISSUE THREAD

**Issue by peak3d** [MetaMask/core#701]
MetaMask Browser extension is one of the most used wallets around.

Fetching fee history introduced by EIP1559 has in most implementations an internal cache of 1024 blocks.
Fetching blocks in a wider range will lead to cache invalidates and huge computations to re-calculate the fee history.

Controller now requests 20.000 past blocks (in ~20 chunks), which runs nearly uncached. This call is done every time someone opens the browser extension, and if I'm not wrong after this in recular n second poll steps.
https://github.com/MetaMask/controllers/blob/main/src/gas/fetchGasEstimatesViaEthFeeHistory/BlockFeeHistoryDatasetFetcher.ts#L55

This means lot of cpu / power consumtion for node owner, gateway providers have to invest heavily in infrastructure to be able to respond to the requests.

Pls. investigate, if it is really required to request the last 20.000 blocks, especially for the MM browser extension I don't think it is required. Using 1024 (cache-size) blocks would give you > 4hours in the past (for ethereum 13 sec block time)

I wouldn't say anything if MM would be a small application used by few, but your success should also come with some responsibility.

**Comment by danfinlay** [comment]
Thanks for the report, we're looking into it!

**Comment by mcmire** [comment]
Hi @peak3d,

Thank you for bringing this up. We appreciate your concern about performance. That said, I wanted to clarify some things about the code that you referenced and how it's used.

> Controller now requests 20.000 past blocks (in ~20 chunks), which runs nearly uncached. This call is done every time someone opens the browser extension

This can happen, but usually does not. We typically we hit an internal API that acts as a proxy to the network. You can see that call [here](https://github.com/MetaMask/controllers/blob/89c00500b9650d5748eae3f6d667f726b9d3cc6a/src/gas/determineGasFeeCalculations.ts#L77), which ends up calling [this](https://github.com/MetaMask/controllers/blob/89c00500b9650d5748eae3f6d667f726b9d3cc6a/src/gas/gas-util.ts#L32) (the URL referenced is located in the extension repo [here](https://github.com/MetaMask/metamask-extension/blob/b8fc1ae671c153d84de6caa254de3e55c153de75/app/scripts/metamask-controller.js#L336)). The API makes several calls to the network — one of which requests 20,000 blocks to support some new information about the network we are now presenting users as a part of the [gas fee UI update](https://twitter.com/MetaMask/status/1498730701939789831) — and then caches the results.

Therefore, the vast majority of the time, the extension works just as you'd expect it to, and we don't blast the network.

The code that you're referring to is a fallback implementation. If for some reason the API is down, then the extension will make calls to the network manually. This logic is designed to mimic what the API is doing as closely as possible so that users are unaware and MetaMask doesn't give them wildly different estimates. That was the original thinking behind this at least.

Now to your point:

> Pls. investigate, if it is really required to request the last 20.000 blocks, especially for the MM browser extension I don't think it is required. Using 1024 (cache-size) blocks would give you > 4hours in the past (for ethereum 13 sec block time)

I do agree that it is probably unnecessary to request so much data in the fallback case. I just talked to the person who usually runs the API I mentioned above and they agreed that this would be a fine change to make. I can get started on making that change right away.

**Comment by tgmichel** [comment]
> The code that you're referring to is a fallback implementation. If for some reason the API is down, then the extension will make calls to the network manually.

Is that also the case for when the API is up but returns an error when there is no data indexed for the given chain id? I.e.

```
{
  "error": "Cannot read property 'getBlock' of undefined"
}
```

**Comment by peak3d** [comment]
> his can happen, but is not likely to happen. We typically we hit an internal API that acts as a proxy to the network.

Thank you for taking time to explain it in detail, and yes, the proxy backing might explain why the topic has not popped up sooner. Appreciate this!

Is the list of registered api endpoints / chains public in MM repository?

- peak3d

**Comment by mcmire** [comment]
Thanks for originally reporting this issue. This is fixed as of MetaMask 10.11.2, which minimizes the number of requests made to the network in the fallback scenario as mentioned above (this includes removal of the 20,000-block request). You see the changes here:

* #705
* #712

Hope this helps. Let us know if you notice anything else!

**Comment by peak3d** [comment]
Thx @mcmire !
